[Train] Extend NCCL RAS hang detection with nvidia-smi and flight recorder (#64993)
ray-project/ray: doc/source/train/user-guides/fault-tolerance.rst

    The workers download the checkpoint from storage and use it to resume training.

.. _train-nccl-hang-detection:

> NCCL Hang Detection
> -------------------

Reviewer: The style guide asks for [sentence case in headings](https://docs.ray.io/en/master/ray-contribute/writing-style.html#headings), so this would be `NCCL hang detection`.

Worth saying plainly that this page doesn't follow that today: every other heading here is title case (`Handling Failures and Node Preemption`, `Worker Process and Node Fault Tolerance`, `Job Driver Fault Tolerance`). So you're choosing between matching the guide and matching the page, and either is reasonable. I'd take the suggestion and let a separate pass fix the rest of the page, but I won't push it.

```suggestion
NCCL hang detection
-------------------
```

> Distributed GPU training can *hang* rather than crash: if ranks disagree on which
> collective to run (a divergent code path, a rank that exited early, or a
> mismatched shape/dtype/call order), NCCL blocks forever waiting for a peer that
> will never arrive. Because no process dies, Ray's normal worker-failure handling
> never triggers and the job sits idle burning GPU hours until a wall-clock timeout.

Reviewer: Two style-guide points in this paragraph, both mechanical:

- The colon after `rather than crash` introduces a clause rather than a list or a code block, which the guide [reserves it for](https://docs.ray.io/en/master/ray-contribute/writing-style.html#sentence-structure).
- `shape/dtype/call order` uses slashes in prose, where the guide asks for [words instead of symbols](https://docs.ray.io/en/master/ray-contribute/writing-style.html#use-words-for-symbols-in-prose).

Splitting the sentence also lets the causes read as the open-ended list they are. I kept `such as` rather than counting them, so this doesn't claim the three are exhaustive.

```suggestion
Distributed GPU training can *hang* rather than crash. When ranks disagree on
which collective to run, NCCL blocks forever waiting for a peer that never
arrives. Ranks disagree for reasons such as a divergent code path, a rank that
exited early, or a mismatched shape, dtype, or call order. Because no process
dies, Ray's normal worker-failure handling never triggers, and the job sits idle
burning GPU hours until a wall-clock timeout.
```

> For runs using the NCCL backend, Ray Train ships an opt-in hang detector,
> enabled by setting ``RAY_TRAIN_ENABLE_NV_HANG_DETECTOR=1`` on the driver.
> It builds on NCCL's `RAS (Reliability, Availability, Serviceability) subsystem
> <https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/ras.html>`_
> (NCCL >= 2.28, via the ``ncclras`` client binary): the controller periodically
> queries each communicator's per-rank collective *launch counts*. A communicator
> whose counts are mismatched **and** make no progress across the confirmation
> window is flagged as hung. On confirmation the detector logs the RAS report and
> the suspected culprit workers (world rank, host, PID, CUDA device — the world
> rank is recovered by matching the RAS-reported host and PID against the worker
> group), captures diagnostics to your storage path, and — depending on the
> configured action — either fails the run with a ``NCCLHangError``, which also
> names the culprits in its ``worker_failures`` keyed by world rank, or only
> observes.

Reviewer: This is the one finding I'd actually like to see fixed. It's a single ~110-word sentence carrying two em-dash asides and a nested parenthetical, and the reader has to hold the whole thing open to get to `or only observes` at the end. The guide asks you to [split rather than join clauses with punctuation, and to avoid dashes and parentheticals in prose](https://docs.ray.io/en/master/ray-contribute/writing-style.html#sentence-structure).

Also in scope here: `NCCL >= 2.28` uses an operator in prose, and the bold on **and** is emphasis in ordinary prose, which the guide [reserves bold against](https://docs.ray.io/en/master/ray-contribute/writing-style.html#bold-and-italics).

On the version: I wrote "the `ncclras` client binary from NCCL 2.28 or later" rather than attaching 2.28 to RAS itself, because the constraint I can see in the code is on the binary's `-f` flag (`nv_hang_detector.py` logs "which requires NCCL 2.28+" when `-f` is missing, and the JSON parser targets the 2.28-2.30 schema). If RAS as a whole also lands in 2.28, say so directly and ignore my hedge.

```suggestion
For runs using the NCCL backend, Ray Train ships an opt-in hang detector. Enable
it by setting ``RAY_TRAIN_ENABLE_NV_HANG_DETECTOR=1`` on the driver. The detector
builds on NCCL's `RAS (Reliability, Availability, Serviceability) subsystem
<https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/ras.html>`_,
querying it through the ``ncclras`` client binary from NCCL 2.28 or later. The
controller periodically queries each communicator's per-rank collective *launch
counts*, and flags any communicator whose counts are mismatched and make no
progress across the confirmation window.

On confirmation, the detector logs the RAS report and the suspected culprit
workers, naming the world rank, host, PID, and CUDA device of each. Ray Train
recovers the world rank by matching the RAS-reported host and PID against the
worker group. The detector captures diagnostics to your storage path, then acts
according to ``RAY_TRAIN_NCCL_RAS_ACTION``: ``fail`` raises a ``NCCLHangError``
that names the culprits in its ``worker_failures``, keyed by world rank, and
``observe`` records the event without failing the run.
```

     - Enable the detector. Default ``0`` (off). Enable it only for NCCL
       jobs, since the RAS subsystem exists only for NCCL.
   * - ``RAY_TRAIN_NCCL_RAS_ACTION``
     - ``observe`` (default) logs diagnostics and captures stacks without
       failing the run; ``fail`` additionally raises ``NCCLHangError`` to
       terminate the run.
   * - ``RAY_TRAIN_NCCL_RAS_POLL_INTERVAL_S``
     - Seconds between RAS polls. Default ``15``.
   * - ``RAY_TRAIN_NCCL_RAS_CONFIRM_WINDOW_S``
     - Seconds a communicator must stay frozen before a hang is confirmed.
       Default ``600`` (10 minutes).
   * - ``RAY_TRAIN_NCCLRAS_PATH``
     - Path to the ``ncclras`` client binary used to query RAS. Default
       ``ncclras``, resolved on the workers' ``PATH``. If the binary is
       missing or predates NCCL 2.28 (no ``-f`` support), the detector logs a
       warning and disables itself for the rest of the run.
   * - ``NCCL_RAS_ADDR``
     - NCCL's own RAS listen address (``host:port``). Read on the *workers*
       when querying, defaulting to ``localhost:28028``; set it in the worker
       environment if your NCCL deployment uses a non-default address.

>    * - ``TORCH_NCCL_TRACE_BUFFER_SIZE``
>      - Opt-in (default off). PyTorch's own flight-recorder knob: set it to a
>        positive ring-buffer size (for example ``20000``) on the driver, and Ray
>        Train propagates it to the workers before the process group is created
>        and dumps the per-rank flight-recorder trace (as JSON) at a confirmed
>        hang for a finer-grained collective timeline.

Reviewer: This cell packs the enable step, the propagation behavior, and the payoff into one sentence hinged on a colon. Splitting it makes the reader's action easier to find. No values changed: still opt-in, still off by default, still `20000` as the example.

```suggestion
   * - ``TORCH_NCCL_TRACE_BUFFER_SIZE``
     - Opt-in, off by default. PyTorch's own flight-recorder knob. Set it on the
       driver to a positive ring-buffer size, for example ``20000``. Ray Train
       propagates it to the workers before it creates the process group, then
       dumps each rank's flight-recorder trace as JSON at a confirmed hang,
       giving you a finer-grained collective timeline.
```

> We recommend rolling out in ``observe`` mode first to establish that the detector
> does not fire on your healthy workloads, then switching to ``fail`` to have hung
> runs terminate instead of hanging indefinitely. ``NCCLHangError`` is terminal:
> because a deterministic hang would usually just hang again on a restart, it is
> not retried by :ref:`worker fault tolerance <train-worker-fault-tolerance>`
> regardless of ``max_failures``. Relaunching the job resumes from the latest
> checkpoint as described in
> :ref:`job driver fault tolerance <train-job-driver-fault-tolerance>`.

Reviewer: Three things here:

- `We recommend rolling out` is [first-person plural](https://docs.ray.io/en/master/ray-contribute/writing-style.html#avoid-first-person-plural), and the guide also asks you to [turn a recommendation into direct advice](https://docs.ray.io/en/master/ray-contribute/writing-style.html#turn-hedging-into-direct-advice). The imperative does both.
- `does not fire` and `it is not retried` want [contractions](https://docs.ray.io/en/master/ray-contribute/writing-style.html#use-contractions).
- `it is not retried by worker fault tolerance` is passive with a nameable actor right there in the sentence.

```suggestion
Roll out in ``observe`` mode first to establish that the detector doesn't fire on
your healthy workloads, then switch to ``fail`` so that hung runs terminate
instead of hanging indefinitely. ``NCCLHangError`` is terminal. A deterministic
hang usually hangs again after a restart, so :ref:`worker fault tolerance
<train-worker-fault-tolerance>` doesn't retry it, regardless of ``max_failures``.
Relaunching the job resumes from the latest checkpoint, as described in
:ref:`job driver fault tolerance <train-job-driver-fault-tolerance>`.
```

.. note::

>     **Known limitations.** RAS tracks collective *launch* counts, so a *symmetric*
>     hang where every rank has launched the same collective and the network or
>     hardware then wedges mid-collective (a switch failure or NIC flap) shows zero
>     skew and is not detected — this needs completion-timing telemetry or the flight
>     recorder. Conversely, pipeline-parallel / point-to-point (``send``/``recv``)
>     communicators can legitimately sit skewed and idle for long stretches; the
>     ``observe`` default plus the time-based confirmation window guard against
>     spurious failures, but consider raising
>     ``RAY_TRAIN_NCCL_RAS_CONFIRM_WINDOW_S`` for such workloads before enabling
>     ``fail``.

Reviewer: Same pattern as the second paragraph: two long sentences, an em-dash aside, a semicolon, and `pipeline-parallel / point-to-point` plus `send`/`recv` as slashes. `consider raising` is also hedged where the guide wants [direct advice](https://docs.ray.io/en/master/ray-contribute/writing-style.html#turn-hedging-into-direct-advice).

I kept the `**Known limitations.**` lead-in, which is exactly the [bold use the guide allows](https://docs.ray.io/en/master/ray-contribute/writing-style.html#bold-and-italics) in an admonition, and kept the switch-failure and NIC-flap examples.

```suggestion
    **Known limitations.** RAS tracks collective *launch* counts, so it misses a
    *symmetric* hang, where every rank has launched the same collective and the
    network or hardware then wedges mid-collective, as in a switch failure or a
    NIC flap. That case shows zero skew, and detecting it needs completion-timing
    telemetry or the flight recorder. Pipeline-parallel and point-to-point
    (``send`` and ``recv``) communicators are the opposite risk. They can
    legitimately sit skewed and idle for long stretches. The ``observe`` default
    and the time-based confirmation window guard against spurious failures on
    those workloads, but raise ``RAY_TRAIN_NCCL_RAS_CONFIRM_WINDOW_S`` for them
    before you enable ``fail``.
```
